backend/utils/pubsub.js

```diff
@@ -0,0 +1,261 @@
+ * {(boolean|object)} perMessageDeflate - Enables/disables per-message deflate.
+ * @returns {Object}
+ */
+export function createServer (httpServer: Object, options?: Object = {}): Object {
+  const server = new WebSocket.Server({
+    ...defaultOptions,
+    ...options,
+    server: httpServer
+  })
+  server.customClientHandlers = { ...options.clientHandlers }
+  server.customServerHandlers = { ...options.serverHandlers }
+  server.messageHandlers = { ...defaultMessageHandlers, ...options.messageHandlers }
+  server.pingInterval = undefined
+  server.subscribersByContractID = Object.create(null)
+
+  // Create and attach server-side WebSocket event listeners.
+  ;['close', 'connection', 'error', 'headers', 'listening'].forEach((name) => {
+    server.on(name, (...args) => {
+      console.log('[pubsub] Server event:', name)
+      try {
+        const customHandler = server.customServerHandlers['on' + name]
+        const defaultHandler = defaultServerHandlers['on' + name]
+        // Always call the default handler first.
+        if (defaultHandler) {
+          defaultHandler.call(server, ...args)
+        }
+        if (customHandler) {
+          customHandler.call(server, ...args)
+        }
+      } catch (error) {
+        logger(error)
+      }
+    })
+  })
+  // Setup a ping interval if required.
+  if (server.options.pingInterval > 0) {
+    server.pingInterval = setInterval(() => {
+      server.clients.forEach((client) => {
+        if (!client.isAlive) {
+          return client.terminate()
+        }
+        client.isAlive = false
+        client.ping()
+      })
+    }, server.options.pingInterval)
+  }
+  return server
+}
+
+const defaultOptions = {
+  clientTracking: true,
+  maxPayload: 6 * 1024 * 1024,
+  pingInterval: undefined
+}
+
+const generateSocketID = (() => {
+  let counter = 0
+
+  return () => counter++
+})()
+
+// ====== Server socket event handlers ====== //
+
+const defaultServerHandlers = Object.create(null)
+
+/**
+ * Emitted when a connection handshake completes.
+ *
+ * @param socket - The client socket that connected.
+ * @param request - The http GET request sent by the socket.
+ */
+defaultServerHandlers.onconnection = function onconnection (socket, request) {
+  socket.id = generateSocketID()
+  socket.isAlive = true
+  socket.server = this
+  socket.subscriptions = new Set()
+
+  const xForwardedFor = request.headers['x-forwarded-for']
+
+  const remoteAddress = (
+    typeof xForwardedFor === 'string' && xForwardedFor !== ''
+      ? xForwardedFor.split(/\s*,\s*/)[0]
+      : request.socket.remoteAddress
+  )
+  console.log(bold(`[pubsub] Socket ${socket.id} connected from ${remoteAddress}`))
+
+  // Create and attach socket event listeners.
+  ;['close', 'error', 'message', 'ping', 'pong'].forEach((eventName) => {
+    socket.on(eventName, (...args) => {
+      console.log(`[pubsub] Event '${eventName}' on socket ${socket.id}`, ...args)
```

Comment:
Should this not be `console.debug` instead?

Reply:
Yes, possibly. I don't precisely understand when to use which logging level, so I generally just use `console.log` everywhere.